### PR TITLE
catcma (GECCO2024)

### PR DESCRIPTION
**Pull request by ha-mano**
I have implemented CatCMA [Hamano+, GECCO2024].

### CODE REVIEW

File: cmaes/_catcma.py [hunk 1]
```diff
@@ -0,0 +1,530 @@
...
+        assert sigma > 0, "sigma must be non-zero positive value"
+
+        assert np.all(
+            np.abs(mean) < _MEAN_MAX
+        ), f"Abs of all elements of mean vector must be less than {_MEAN_MAX}"
+
+        self._n_co = len(mean)
+        self._n_ca = len(cat_num)
+        self._n = self._n_co + self._n_ca
+        assert self._n_co > 1, "The dimension of mean must be larger than 1"
```
Comment (nomuramasahir0) [inline cmaes/_catcma.py:112]: How about adding assert for `self._n_ca` ?

File: cmaes/_catcma.py [hunk 2]
```diff
@@ -0,0 +1,530 @@
...
+        self,
+        mean: np.ndarray,
+        sigma: float,
+        cat_num: np.ndarray,
+        bounds: Optional[np.ndarray] = None,
+        n_max_resampling: int = 100,
+        seed: Optional[int] = None,
+        population_size: Optional[int] = None,
+        cov: Optional[np.ndarray] = None,
+        cat_param: Optional[np.ndarray] = None,
```
Comment (nomuramasahir0) [inline cmaes/_catcma.py:97]: Please add the validation.

File: cmaes/_catcma.py [hunk 3]
```diff
@@ -0,0 +1,530 @@
...
+from typing import cast
+from typing import Optional
+
+
+_EPS = 1e-8
+_MEAN_MAX = 1e32
+_SIGMA_MAX = 1e32
+
+
+class CatCMA:
```
[the remark this answers is not in the document]
Comment (ha-mano) [reply cmaes/_catcma.py:16]: I added the link for the paper (arXiv).

File: cmaes/_catcma.py [hunk 4]
```diff
@@ -0,0 +1,530 @@
...
+        cat_param:
+            A parameter of categorical distribution (optional).
+
+        margin:
+            A margin (lower bound) of categorical distribution (optional).
+
+        min_eigenvalue:
+            Lower bound of eigenvalue of multivariate Gaussian distribution (optional).
+
+        delta_init:
```
[the remark this answers is not in the document]
Comment (ha-mano) [reply cmaes/_catcma.py]: The variables `delta_init` and `Delta_max` have been removed from the `__init__` method.

File: cmaes/_catcma.py [hunk 5]
```diff
@@ -0,0 +1,530 @@
...
+            Initial trust region radius in ASNG (optional).
+
+        Delta_max:
```
[the remark this answers is not in the document]
Comment (ha-mano) [reply cmaes/_catcma.py]: The variables `delta_init` and `Delta_max` have been removed from the `__init__` method (as mentioned above).

**Comment by nomuramasahir0** [comment]
Thanks! LGTM.

**Pull request merged**